cps/kni.c

```diff
@@ -613,10 +661,138 @@ mnl_socket_recvfrom_flags(const struct mnl_socket *nl, void *buf, size_t bufsiz,
+
+	if (update->family == AF_INET6) {
+		proto = ETHER_TYPE_IPv6;
+		gw_fib_id = lpm_lookup_ipv6(ltbl->lpm6, update->gw.v6.s6_addr);
+		if (gw_fib_id < 0)
+			return -1;
+		gw_fib = &ltbl->fib_tbl6[gw_fib_id];
+
+		if (inet_ntop(AF_INET6, &update->ip.v6.s6_addr,
+				ip_buf, sizeof(ip_buf)) == NULL)
+			return -1;
+
+		ret = snprintf(ipp_buf, sizeof(ipp_buf), "%s/%hhu",
+			ip_buf, update->prefix_len);
+		RTE_VERIFY(ret > 0 && ret < (int)sizeof(ipp_buf));
+	} else {
+		proto = ETHER_TYPE_IPv4;
+		gw_fib_id = lpm_lookup_ipv4(ltbl->lpm, update->gw.v4.s_addr);
+		if (gw_fib_id < 0)
+			return -1;
+		gw_fib = &ltbl->fib_tbl[gw_fib_id];
+
+		if (inet_ntop(AF_INET, &update->ip.v4.s_addr,
+				ip_buf, sizeof(ip_buf)) == NULL)
+			return -1;
+
+		ret = snprintf(ipp_buf, sizeof(ipp_buf), "%s/%hhu",
+			ip_buf, update->prefix_len);
+		RTE_VERIFY(ret > 0 && ret < (int)sizeof(ipp_buf));
+	}
+
+	prefix_info.str = ipp_buf;
+	prefix_info.addr.proto = proto;
+	rte_memcpy(&prefix_info.addr.ip, &update->ip,
+		sizeof(prefix_info.addr.ip));
+	prefix_info.len = update->prefix_len;
+
+	gw_addr.proto = proto;
+	rte_memcpy(&gw_addr.ip, &update->gw, sizeof(gw_addr.ip));
+
+	if (gw_fib->action == GK_FWD_NEIGHBOR_FRONT_NET) {
+		if (update->oif_index != 0) {
+			const char *kni_name =
+				rte_kni_get_name(cps_conf->front_kni);
+			if (update->oif_index != if_nametoindex(kni_name)) {
+				RTE_LOG(WARNING, GATEKEEPER,
+					"cps: the output KNI interface for prefix %s is not the front interface while the gateway for the prefix in Gatekeeper is a neighbor of the front network!\n",
+					prefix_info.str);
+				return -1;
+			}
+		}
+
+		return add_fib_entry_numerical(&prefix_info, NULL, &gw_addr,
+			GK_FWD_GATEWAY_FRONT_NET, cps_conf->gk);
+	}
+
+	if (gw_fib->action == GK_FWD_NEIGHBOR_BACK_NET) {
+		if (update->oif_index != 0) {
+			const char *kni_name =
+				rte_kni_get_name(cps_conf->back_kni);
+			if (update->oif_index != if_nametoindex(kni_name)) {
+				RTE_LOG(WARNING, GATEKEEPER,
+					"cps: the output KNI interface for prefix %s is not the back interface while the gateway for the prefix in Gatekeeper is a neighbor of the back network!\n",
+					prefix_info.str);
+				return -1;
+			}
+		}
+
+		return add_fib_entry_numerical(&prefix_info, NULL, &gw_addr,
+			GK_FWD_GATEWAY_BACK_NET, cps_conf->gk);
+	}
+
+	return -1;
+}
+
+static int
+del_route(struct route_update *update, struct gk_config *gk_conf)
+{
+	int ret;
+	char ip_buf[INET6_ADDRSTRLEN];
+	char ipp_buf[INET6_ADDRSTRLEN + 4];
+	struct ip_prefix prefix_info;
+
+	if (update->family == AF_INET6) {
+		if (inet_ntop(AF_INET6, &update->ip.v6.s6_addr,
+				ip_buf, sizeof(ip_buf)) == NULL)
+			return -1;
+
+		ret = snprintf(ipp_buf, sizeof(ipp_buf), "%s/%hhu",
+			ip_buf, update->prefix_len);
+		RTE_VERIFY(ret > 0 && ret < (int)sizeof(ipp_buf));
+
+		prefix_info.addr.proto = ETHER_TYPE_IPv6;
+		rte_memcpy(&prefix_info.addr.ip.v6, &update->ip.v6,
+			sizeof(prefix_info.addr.ip.v6));
+	} else {
```

Comment:
Let's put IPv4 first, and use the structure `else if (likely(update->family == AF_INET6)) ... else { ...`